ferrocene/doc/internal-procedures/src/working-with-the-ci.rst

```diff
@@ -0,0 +1,250 @@
+.. SPDX-License-Identifier: MIT OR Apache-2.0
+   SPDX-FileCopyrightText: The Ferrocene Developers
+
+Working with the CI
+===================
+
+Ferrocene's Continuous Integration is primarily driven by a combination of
+YAML, Docker, Python and Bash scripts.
+
+
+Imitating the CI
+----------------
+
+Sometimes it is useful to imitate the CI to reproduce failures.
+
+The Linux jobs on CI typically run in containers and can be almost exactly
+reproduced. Mac and Windows jobs on CI cannot run in containers and are more
+difficult to reproduce.
+
+If you are presented the option, it's a good idea to reproduce Linux CI
+failures first.
+
+
+Installing Docker (with Buildkit)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You do not need to download or use `Docker Desktop
+<https://www.docker.com/products/docker-desktop/>`_ with Ferrocene. If you
+choose to use Docker Desktop, install it as directed on the Website and skip
+the rest of this section.
+
+On Linux, install Docker and ``docker-buildx`` from your package manager. On
+Ubuntu 20.04 and up:
+
+.. code-block:: bash
+
+   apt install docker.io docker-buildx
+   
+On some distributions, you may need to start the service with
+``systemctl enable --now docker``. If you just installed ``docker-buildx``,
+restart the Docker service.
+
+On MacOS, you can use the Docker CLI connected to a Lima host:
+
+.. code-block:: bash
+
+   brew install lima docker
+   limactl start template://docker
+   docker context create lima-docker --docker "host=unix://{{.Dir}}/sock/docker.sock"
+   docker context use lima-docker
+   docker run --rm hello-world
```

Comment:
needs a DOCKER_HOST set so as to know where the socket is

did so, with the fish

    set -x DOCKER_HOST (limactl list docker --format 'unix://{{.Dir}}/sock/docker.sock')

Reply:
On my machine `docker context create lima-docker --docker "host=unix://{{.Dir}}/sock/docker.sock"` was sufficient, did it not work for you?

Reply:
did not work for me
```
❯ docker context create lima-docker --docker 'host=unix://{{.Dir}}/sock/docker.sock'
unable to create docker endpoint config: invalid docker endpoint options: parse "unix://{{.Dir}}/sock/docker.sock": invalid character "{" in host name
```

so I replaced `{{.Dir}}` with a literal path

I then ran the 2 other commands with success, then the third failed
```
❯ docker run --rm hello-world
docker: Cannot connect to the Docker daemon at unix://Users/tshepang/.lima/docker/sock/docker.sock. Is the docker daemon running?.
```

that's where I (thought I had to) set `z`, and it worked

---

what helped is using this (long) command

```
❯ docker context create lima-docker --docker host=(limactl list docker --format unix://{{.Dir}}/sock/docker.sock)
lima-docker
Successfully created context "lima-docker"
```
no need for the `DOCKER_HOST` that way

Reply:
This was puzzling, but I did figure it out. 😅

So lima actually templates it the path, so we need to just tell them to refer to the instructions that Lima outputs at the end:

https://github.com/lima-vm/lima/blob/118c249cc8686d65507b9249e931a7feed615497/examples/docker.yaml#L81-L86

This way, they should have the right path.

Regarding `DOCKER_HOST`, the Docker context thing should make that unnecessary?

My local Mac does not have `DOCKER_HOST` set:

```
$ echo $env.DOCKER_HOST
Error: nu::shell::column_not_found

  × Cannot find column 'DOCKER_HOST'
   ╭─[entry #7:1:6]
 1 │ echo $env.DOCKER_HOST
   ·      ──┬─ ─────┬─────
   ·        │       ╰── cannot find column 'DOCKER_HOST'
   ·        ╰── value originates here
   ╰────

Exit status 1

ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ limactl list
NAME       STATUS     SSH                VMTYPE    ARCH       CPUS    MEMORY    DISK      DIR
default    Stopped    127.0.0.1:0        qemu      aarch64    6       16GiB     100GiB    ~/.lima/default
docker     Running    127.0.0.1:59566    vz        aarch64    4       4GiB      100GiB    ~/.lima/docker
riscv      Stopped    127.0.0.1:0        qemu      riscv64    6       16GiB     100GiB    ~/.lima/riscv

$ docker context ls
NAME            DESCRIPTION                               DOCKER ENDPOINT                                   ERROR
default         Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                       
lima-docker *                                             unix:///Users/ana/.lima/docker/sock/docker.sock   

ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ docker context inspect lima-docker
[
    {
        "Name": "lima-docker",
        "Metadata": {},
        "Endpoints": {
            "docker": {
                "Host": "unix:///Users/ana/.lima/docker/sock/docker.sock",
                "SkipTLSVerify": false
            }
        },
        "TLSMaterial": {},
        "Storage": {
            "MetadataPath": "/Users/ana/.docker/contexts/meta/3384fc6800de4f1211539c4930cbe1f5a8229c0eaa8d4241fe1bd7180b0da0e6",
            "TLSPath": "/Users/ana/.docker/contexts/tls/3384fc6800de4f1211539c4930cbe1f5a8229c0eaa8d4241fe1bd7180b0da0e6"
        }
    }
]
```

You can see I can create a new context and start using it:

```
ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ docker context create test-lima-docker --docker "host=unix:///Users/ana/.lima/docker/sock/docker.sock"
test-lima-docker
Successfully created context "test-lima-docker"
ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ docker context use test-lima-docker
test-lima-docker
Current context is now "test-lima-docker"
ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
ana@laniakea.local (Darwin) in git/ferrocene/ferrocene hoverbear/host-aarch64-unknown-linux-gnu (clean)
$ 
```

I wonder if this is related to the version of Docker or Lima you have?

Here's mine:

```
$ lima --version
limactl version 0.22.0

$ docker --version
Docker version 27.0.2, build 912c1ddf8a
```

Reply:
you may have missed the part where I found how to not need DOCKER_HOST
```
❯ docker context create lima-docker --docker host=(limactl list docker --format unix://{{.Dir}}/sock/docker.sock)
```
my versions...
```
❯ lima --version
limactl version 0.22.0

❯ docker --version
Docker version 27.0.3, build 7d4bcd863a

Reply:
Wooo! I did kind of miss that!